src/EVM/SMT.hs

```diff
@@ -221,9 +216,44 @@ findStorageReads = foldProp go []
   where
     go :: Expr a -> [(Expr EWord, Expr EWord)]
     go = \case
-      SLoad addr slot storage -> [(addr, slot) | isAbstractStorage storage]
+      SLoad addr slot storage -> [(addr, slot) | containsNode isAbstractStore storage]
       _ -> []
 
+    isAbstractStore AbstractStore = True
+    isAbstractStore _ = False
+
+
+findBufferAccess :: TraversableTerm a => [a] -> [(Expr EWord, Expr EWord, Expr Buf)]
+findBufferAccess = foldl (\acc p -> foldTerm go acc p) mempty
+  where
+    go :: Expr a -> [(Expr EWord, Expr EWord, Expr Buf)]
+    go = \case
+      ReadWord idx buf -> [(idx, Lit 32, buf)]
+      ReadByte idx buf -> [(idx, Lit 1, buf)]
+      CopySlice srcOff _ size src _  -> [(srcOff, size, src)]
+      _ -> mempty
+
+
+-- | Asserts that buffer reads beyond the size of the buffer are equal
+-- to zero. Looks for buffer reads in the a list of given predicates
+-- and the buffer and storage environments.
+assertReads :: [Prop] -> BufEnv -> StoreEnv -> [Prop]
+assertReads props benv senv = concat $ fmap assertRead allReads
+  where
+    assertRead :: (Expr EWord, Expr EWord, Expr Buf) -> [Prop]
+    assertRead (idx, Lit 32, buf) = [POr (PEq (ReadWord idx buf) (Lit 0)) (PNeg (PGEq idx (bufLength buf)))]
+    assertRead (idx, Lit sz, buf) = fmap (\s -> POr (PEq (ReadByte idx buf) (LitByte (num s))) (PNeg (PGEq idx (bufLength buf)))) [(0::Int)..num sz-1]
+    assertRead (_, _, _) = [] -- cannot generate assertions for accesses of symbolic size
```

Comment:
Is it sound to ignore this case? Should we be erroring out instead?

Reply:
I'd say that it's sound but incomplete. If there is a copySlice with symbolic size, we will not be able to prove that some accesses are zero (however, we will not prove something that is not true, so in this sense it is not unsound). I'm fine with throwing an error here instead though.

I think this is the only case that this will show up is`copySlice` with symbolic size. This is unimplemented in SMT.hs so it will throw an error there anyway. Also, I think that such cases cause an infinite loop in symbolic execution.

Reply:
hmmmm yeah all good points, I guess I kinda feel like it's good to error out here just so that we kinda have a lil todo list for when we do start supporting fully dynamic data

Reply:
Done!